File: aws_batch_squared_py/nextflow_batch_stack.py
"""Top-level orchestration stack.

Ported from ``lib/nextflow-batch-stack.ts``. Wires all nested stacks together,
publishes queue/bucket references to SSM (so the Nextflow head-node role can
resolve them at deploy time), and declares cross-stack dependencies.
"""

import aws_cdk as cdk
from aws_cdk import aws_ssm as ssm
from constructs import Construct

from .config import NextflowBatchConfig
from .batch_stack import BatchStack
from .iam_stack import IamStack
from .launch_template_stack import LaunchTemplateStack
from .nextflow_ecr_stack import NextflowEcrStack
from .nextflow_stack import NextflowStack
from .s3_stack import S3Stack
from .vpc_stack import VpcStack


class NextflowBatchStack(cdk.Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        config: NextflowBatchConfig,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Create VPC Stack (or use existing)
        vpc_stack = VpcStack(
            self,
            "VpcStack",
            namespace=config.namespace,
            create_vpc=config.create_vpc,
            existing_vpc_id=config.vpc_id,
            existing_subnet_ids=config.subnet_ids,
        )

        # Create S3 Stack
        s3_stack = S3Stack(
            self,
            "S3Stack",
            namespace=config.namespace,
            group_name=config.group_name,
            bucket_name=config.s3_bucket_name,
            existing_bucket=config.existing_bucket,
            work_dir_expiration_days=config.work_dir_expiration_days,
        )

        # Create IAM Stack
        iam_stack = IamStack(
            self,
            "IamStack",
            s3_bucket_name=s3_stack.bucket_name,
        )
        iam_stack.add_dependency(s3_stack)

        # Create Launch Template Stack
        launch_template_stack = LaunchTemplateStack(
            self,
            "LaunchTemplateStack",
            namespace=config.namespace,
            group_name=config.group_name,
            batch_compute_ami=config.batch_compute_ami,
            s3_bucket_name=s3_stack.bucket_name,
            s3_reference_path=config.s3_reference_path,
        )

        # Create Batch Stack
        batch_stack = BatchStack(
            self,
            "BatchStack",
            namespace=config.namespace,
            vpc=vpc_stack.vpc,
            subnets=vpc_stack.subnets,
            launch_template_id=launch_template_stack.launch_template_id,
            batch_service_role=iam_stack.batch_service_role,
            instance_profile=iam_stack.batch_instance_profile,
            spot_fleet_role=iam_stack.batch_spot_fleet_role,
            on_demand_min_cpus=config.on_demand_min_cpus,
            on_demand_max_cpus=config.on_demand_max_cpus,
            spot_min_cpus=config.spot_min_cpus,
            spot_max_cpus=config.spot_max_cpus,
            batch_on_demand_instance_types=config.batch_on_demand_instance_types,
            batch_spot_instance_types=config.batch_spot_instance_types,
        )
        batch_stack.add_dependency(iam_stack)
        batch_stack.add_dependency(launch_template_stack)
        batch_stack.add_dependency(vpc_stack)

        # Build Nextflow image if needed
        if config.build_nextflow_image:
            nextflow_ecr_stack = NextflowEcrStack(
                self,
                "NextflowEcrStack",
                namespace=config.namespace,
            )
            nextflow_image_uri = nextflow_ecr_stack.image_uri
        else:
            if not config.existing_nextflow_image:
                raise ValueError(
                    "existingNextflowImage must be provided when buildNextflowImage is false"
                )
            nextflow_image_uri = config.existing_nextflow_image

        # Create SSM Parameters
        s3_bucket_param = ssm.StringParameter(
            self,
            "ParamS3Bucket",
            parameter_name=f"/{config.group_name}/{config.namespace}/s3-bucket",
            string_value=s3_stack.bucket_name,
            description="S3 Bucket",
        )

        on_demand_queue_param = ssm.StringParameter(
            self,
            "ParamOnDemandJobQueue",
            parameter_name=f"/{config.group_name}/{config.namespace}/job-queue/on-demand",
            string_value=batch_stack.on_demand_queue_arn,
            description="On-Demand AWS Batch Job Queue",
        )

        spot_queue_param = ssm.StringParameter(
            self,
            "ParamSpotJobQueue",
            parameter_name=f"/{config.group_name}/{config.namespace}/job-queue/spot",
            string_value=batch_stack.spot_queue_arn,
            description="Spot AWS Batch Job Queue",
        )

        # Create Nextflow Stack
        nextflow_stack = NextflowStack(
            self,
            "NextflowStack",
            namespace=config.namespace,
            group_name=config.group_name,
            nextflow_image=nextflow_image_uri,
            s3_bucket_param=s3_bucket_param.parameter_name,
            on_demand_queue_param=on_demand_queue_param.parameter_name,
            spot_queue_param=spot_queue_param.parameter_name,
        )
        nextflow_stack.add_dependency(batch_stack)
        nextflow_stack.node.add_dependency(s3_bucket_param)
        nextflow_stack.node.add_dependency(on_demand_queue_param)
        nextflow_stack.node.add_dependency(spot_queue_param)

        # Outputs
        cdk.CfnOutput(self, "S3BucketName", value=s3_stack.bucket_name)
        cdk.CfnOutput(self, "BatchJobRoleArn", value=iam_stack.batch_job_role.role_arn)
        cdk.CfnOutput(
            self, "OnDemandJobQueueArn", value=batch_stack.on_demand_queue_arn
        )
        cdk.CfnOutput(self, "SpotJobQueueArn", value=batch_stack.spot_queue_arn)

File: playbook/ways-to-run/patterns/batch-squared/infrastructure/lib/nextflow-batch-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';
import { VpcStack } from './vpc-stack';
import { S3Stack } from './s3-stack';
import { IamStack } from './iam-stack';
import { LaunchTemplateStack } from './launch-template-stack';
import { BatchStack } from './batch-stack';
import { NextflowStack } from './nextflow-stack';
import { NextflowEcrStack } from './nextflow-ecr-stack';

export interface NextflowBatchConfig {
  namespace: string;
  groupName: string;
  createVpc: boolean;
  vpcId?: string;
  subnetIds?: string[];
  s3BucketName?: string;
  existingBucket: boolean;
  buildNextflowImage: boolean;
  existingNextflowImage?: string;
  batchComputeAmi?: string;
  s3ReferencePath?: string;
  onDemandMinCpus?: number;
  onDemandMaxCpus?: number;
  spotMinCpus?: number;
  spotMaxCpus?: number;
  batchOnDemandInstanceTypes?: string;
  batchSpotInstanceTypes?: string;
  workDirExpirationDays?: number;
}

export interface NextflowBatchStackProps extends cdk.StackProps {
  config: NextflowBatchConfig;
}

export class NextflowBatchStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: NextflowBatchStackProps) {
    super(scope, id, props);

    const { config } = props;

    // Create VPC Stack (or use existing)
    const vpcStack = new VpcStack(this, 'VpcStack', {
      namespace: config.namespace,
      createVpc: config.createVpc,
      existingVpcId: config.vpcId,
      existingSubnetIds: config.subnetIds,
    });

    // Create S3 Stack
    const s3Stack = new S3Stack(this, 'S3Stack', {
      namespace: config.namespace,
      groupName: config.groupName,
      bucketName: config.s3BucketName,
      existingBucket: config.existingBucket,
      workDirExpirationDays: config.workDirExpirationDays,
    });

    // Create IAM Stack
    const iamStack = new IamStack(this, 'IamStack', {
      s3BucketName: s3Stack.bucketName,
    });
    iamStack.addDependency(s3Stack);

    // Create Launch Template Stack
    const launchTemplateStack = new LaunchTemplateStack(this, 'LaunchTemplateStack', {
      namespace: config.namespace,
      groupName: config.groupName,
      batchComputeAmi: config.batchComputeAmi!,
      s3BucketName: s3Stack.bucketName,
      s3ReferencePath: config.s3ReferencePath!,
    });

    // Create Batch Stack
    const batchStack = new BatchStack(this, 'BatchStack', {
      namespace: config.namespace,
      vpc: vpcStack.vpc,
      subnets: vpcStack.subnets,
      launchTemplateId: launchTemplateStack.launchTemplateId,
      batchServiceRole: iamStack.batchServiceRole,
      instanceProfile: iamStack.batchInstanceProfile,
      spotFleetRole: iamStack.batchSpotFleetRole,
      onDemandMinCpus: config.onDemandMinCpus!,
      onDemandMaxCpus: config.onDemandMaxCpus!,
      spotMinCpus: config.spotMinCpus!,
      spotMaxCpus: config.spotMaxCpus!,
      batchOnDemandInstanceTypes: config.batchOnDemandInstanceTypes!,
      batchSpotInstanceTypes: config.batchSpotInstanceTypes!,
    });
    batchStack.addDependency(iamStack);
    batchStack.addDependency(launchTemplateStack);
    batchStack.addDependency(vpcStack);

    // Build Nextflow image if needed
    let nextflowImageUri: string;
    if (config.buildNextflowImage) {
      const nextflowEcrStack = new NextflowEcrStack(this, 'NextflowEcrStack', {
        namespace: config.namespace,
      });
      nextflowImageUri = nextflowEcrStack.imageUri;
    } else {
      if (!config.existingNextflowImage) {
        throw new Error('existingNextflowImage must be provided when buildNextflowImage is false');
      }
      nextflowImageUri = config.existingNextflowImage;
    }

    // Create SSM Parameters
    const s3BucketParam = new ssm.StringParameter(this, 'ParamS3Bucket', {
      parameterName: `/${config.groupName}/${config.namespace}/s3-bucket`,
      stringValue: s3Stack.bucketName,
      description: 'S3 Bucket',
    });

    const onDemandQueueParam = new ssm.StringParameter(this, 'ParamOnDemandJobQueue', {
      parameterName: `/${config.groupName}/${config.namespace}/job-queue/on-demand`,
      stringValue: batchStack.onDemandQueueArn,
      description: 'On-Demand AWS Batch Job Queue',
    });

    const spotQueueParam = new ssm.StringParameter(this, 'ParamSpotJobQueue', {
      parameterName: `/${config.groupName}/${config.namespace}/job-queue/spot`,
      stringValue: batchStack.spotQueueArn,
      description: 'Spot AWS Batch Job Queue',
    });

    // Create Nextflow Stack
    const nextflowStack = new NextflowStack(this, 'NextflowStack', {
      namespace: config.namespace,
      groupName: config.groupName,
      nextflowImage: nextflowImageUri,
      s3BucketParam: s3BucketParam.parameterName,
      onDemandQueueParam: onDemandQueueParam.parameterName,
      spotQueueParam: spotQueueParam.parameterName,
    });
    nextflowStack.addDependency(batchStack);
    nextflowStack.node.addDependency(s3BucketParam);
    nextflowStack.node.addDependency(onDemandQueueParam);
    nextflowStack.node.addDependency(spotQueueParam);

    // Outputs
    new cdk.CfnOutput(this, 'S3BucketName', {
      value: s3Stack.bucketName,
    });

    new cdk.CfnOutput(this, 'BatchJobRoleArn', {
      value: iamStack.batchJobRole.roleArn,
    });

    new cdk.CfnOutput(this, 'OnDemandJobQueueArn', {
      value: batchStack.onDemandQueueArn,
    });

    new cdk.CfnOutput(this, 'SpotJobQueueArn', {
      value: batchStack.spotQueueArn,
    });
  }
}